src/backend/columnar/columnar_storage.c

```diff
@@ -0,0 +1,281 @@
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *)metapage, sizeof(ColumnarMetapage));
+}
+
+
+/*
+ * ColumnarMetapageReserveData - reserve logical data offsets for writing.
+ */
+uint64
+ColumnarMetapageReserveData(Relation rel, uint64 amount)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
+
+	ColumnarMetapage metapage;
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *) &metapage, sizeof(ColumnarMetapage));
+
+	/* make reservation starting on new page */
+	SmgrAddr prev_addr = logical_to_smgr(metapage.reserved_offset);
+	SmgrAddr initial = { 0 };
+	if (prev_addr.offset == SizeOfPageHeaderData)
+	{
+		initial.blockno = prev_addr.blockno;
+	}
+	else
+	{
+		initial.blockno = prev_addr.blockno + 1;
+	}
+	initial.offset = SizeOfPageHeaderData;
+
+	uint64 reservation_start = smgr_to_logical(initial);
+	metapage.reserved_offset = reservation_start + amount;
+
+	SmgrAddr final = logical_to_smgr(metapage.reserved_offset);
+
+	/* extend with new pages */
+	RelationOpenSmgr(rel);
+	uint64 nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
+
+	while (nblocks <= final.blockno)
+	{
+		Buffer newBuffer = ReadBuffer(rel, P_NEW);
+		Assert(BufferGetBlockNumber(newBuffer) == nblocks);
+		ReleaseBuffer(newBuffer);
+		nblocks++;
+	}
+
+	/* write new reservation */
+	WriteToBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				 (char *) &metapage, sizeof(ColumnarMetapage), true);
+
+	UnlockRelationForExtension(rel, ExclusiveLock);
+
+	return reservation_start;
+}
+
+
+/*
+ * ColumnarMetapageReserveData - reserve stripe ID and row numbers.
+ */
+uint64
+ColumnarMetapageReserveStripe(Relation rel, uint64 nrows, uint64 *first_row)
+{
+	LockRelationForExtension(rel, ExclusiveLock);
+
+	ColumnarMetapage metapage;
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+	ReadFromBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				  (char *) &metapage, sizeof(ColumnarMetapage));
+
+	uint64 stripe_id = metapage.reserved_stripe_id;
+	metapage.reserved_stripe_id++;
+
+	*first_row = metapage.reserved_row_number;
+	metapage.reserved_row_number += nrows;
+
+	WriteToBlock(rel, COLUMNAR_METAPAGE_BLOCKNO, SizeOfPageHeaderData,
+				 (char *) &metapage, sizeof(ColumnarMetapage), true);
+
+	UnlockRelationForExtension(rel, ExclusiveLock);
+
+	return stripe_id;
+}
+
+void
+ColumnarStorageRead(Relation rel, uint64 logicalOffset, char *data, uint32 amount)
+{
+	uint64 read = 0;
+
+	Assert(LogicalOffsetIsValid(logicalOffset));
+
+	while (read < amount)
+	{
+		SmgrAddr addr = logical_to_smgr(logicalOffset + read);
+
+		uint32 to_read = Min(amount - read, BLCKSZ - addr.offset);
+		ReadFromBlock(rel, addr.blockno, addr.offset, data + read, to_read);
+
+		read += to_read;
+	}
+}
+
+
+void
+ColumnarStorageWrite(Relation rel, uint64 logicalOffset, char *data, uint32 amount)
```

Comment:
missing comment

(btw, thanks for breaking `WriteToSmgr` logic into `ColumnarStorageWrite` & `WriteToBlock`, looks much more readable)